File: daily_problems/2024/08/0820/personal_submission/cf1076d_yawn_sean.py
# Submission link: https://codeforces.com/contest/1076/submission/277242827
def main():
    n, m, k = MII()

    us = []
    vs = []
    ws = []

    path = [[] for _ in range(n)]

    for i in range(m):
        u, v, w = MII()
        u -= 1
        v -= 1
        us.append(u)
        vs.append(v)
        ws.append(w)
        path[u].append(i)
        path[v].append(i)

    dis = [10 ** 15] * n
    dis[0] = 0

    parent = [-1] * n
    hpq = [0]

    while hpq:
        d, u = divmod(heappop(hpq), n)
        if dis[u] == d:
            for i in path[u]:
                v = us[i] ^ vs[i] ^ u
                if dis[v] > dis[u] + ws[i]:
                    dis[v] = dis[u] + ws[i]
                    parent[v] = i
                    heappush(hpq, dis[v] * n + v)

    st_range = sorted(range(1, n), key=lambda x: dis[x])
    ans = min(k, n - 1)
    print(ans)
    print(' '.join(str(parent[st_range[i]] + 1) for i in range(ans)))

File: daily_problems/2024/08/0820/personal_submission/cf1076d_yawn_sean.cpp
#include <bits/stdc++.h>
// #pragma GCC optimize("O3,Ofast,unroll-loops")
// #include <bits/extc++.h>
// using namespace __gnu_cxx;
// using namespace __gnu_pbds;
using namespace std;

signed main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);
    
    int n, m, k;
    cin >> n >> m >> k;

    vector<int> us(m), vs(m), ws(m);
    vector<vector<int>> path(n);

    for (int i = 0; i < m; i ++) {
        cin >> us[i] >> vs[i] >> ws[i];
        us[i] --, vs[i] --;
        path[us[i]].emplace_back(i);
        path[vs[i]].emplace_back(i);
    }

    vector<long long> dis(n, 1e15);
    dis[0] = 0;

    vector<int> parent(n, -1);
    priority_queue<pair<long long, int>, vector<pair<long long, int>>, greater<pair<long long, int>>> pq;
    pq.push({0, 0});

    while (pq.size()) {
        auto [d, u] = pq.top();
        pq.pop();
        if (dis[u] == d) {
            for (auto &i: path[u]) {
                int v = us[i] ^ vs[i] ^ u;
                if (dis[v] > dis[u] + ws[i]) {
                    dis[v] = dis[u] + ws[i];
                    parent[v] = i;
                    pq.push({dis[v], v});
                }
            }
        }
    }

    vector<int> st_range(n - 1);
    iota(st_range.begin(), st_range.end(), 1);
    sort(st_range.begin(), st_range.end(), [&] (int i, int j) {return dis[i] < dis[j];});

    int ans = min(k, n - 1);
    cout << ans << '\n';
    for (int i = 0; i < ans; i ++)
        cout << parent[st_range[i]] + 1 << ' ';

    return 0;
}